src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -4,274 +4,213 @@
         {
             options = options ?? new Options();
-            try
-            {
-                var server = new Socket(options.Address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
-
-                server.Bind(new IPEndPoint(options.Address, 0));
-                server.Listen(options.ListenBacklog);
 
-                localEndPoint = (IPEndPoint)server.LocalEndPoint;
-                string host = options.Address.AddressFamily == AddressFamily.InterNetworkV6 ? 
-                    $"[{localEndPoint.Address}]" :
-                    localEndPoint.Address.ToString();
+            using (var listenSocket = new Socket(options.Address.AddressFamily, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listenSocket.Bind(new IPEndPoint(options.Address, 0));
+                listenSocket.Listen(options.ListenBacklog);
 
-                string scheme = options.UseSsl ? "https" : "http";
-                if (options.WebSocketEndpoint)
+                using (var server = new LoopbackServer(listenSocket, options))
                 {
-                    scheme = options.UseSsl ? "wss" : "ws";
+                    await funcAsync(server);
                 }
-
-                var url = new Uri($"{scheme}://{host}:{localEndPoint.Port}/");
-
-                return funcAsync(server, url).ContinueWith(t =>
-                {
-                    server.Dispose();
-                    t.GetAwaiter().GetResult();
-                }, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
-            }
-            catch (Exception e)
-            {
-                localEndPoint = null;
-                return Task.FromException(e);
             }
         }
 
-        public static Task CreateServerAndClientAsync(Func<Uri, Task> clientFunc, Func<Socket, Task> serverFunc)
+        public static Task CreateServerAsync(Func<LoopbackServer, Uri, Task> funcAsync, Options options = null)
         {
-            return CreateServerAsync(async (server, uri) =>
+            return CreateServerAsync(server => funcAsync(server, server.Uri), options);
+        }
+
+        public static Task CreateClientAndServerAsync(Func<Uri, Task> clientFunc, Func<LoopbackServer, Task> serverFunc)
+        {
+            return CreateServerAsync(async server =>
             {
-                Task clientTask = clientFunc(uri);
+                Task clientTask = clientFunc(server.Uri);
                 Task serverTask = serverFunc(server);
 
                 await new Task[] { clientTask, serverTask }.WhenAllOrAnyFailed();
             });
         }
 
-        public static string DefaultHttpResponse => $"HTTP/1.1 200 OK\r\nDate: {DateTimeOffset.UtcNow:R}\r\nContent-Length: 0\r\n\r\n";
+        private static string GetDefaultHttpResponse() => $"HTTP/1.1 200 OK\r\nDate: {DateTimeOffset.UtcNow:R}\r\nContent-Length: 0\r\n\r\n";
```

Comment:
Nit: with this being a method now, consider adding a `int contentLength = 0` optional argument and then using that instead `{contentLength}` instead of `0` in the body.  I know there are a few tests where we've avoided using this default because we need to send a response body, and that would allow us to reuse this.

Reply:
Yeah, I'll look at this.

Reply:
Looking at the various usage of LoopbackServer, most non-default responses do one of the following:

(1) Default response with string content -- could add GetDefaultHttpResponseWithContent(string content), this would add correct Content-Length as well
(2) Default response with added header (e.g. Location) -- could add an overload that allows specifying additional headers
(3) Default response with modified status code, e.g. 302 or 401 -- could add an overload for this as well

Or some combination of the above.  I'm leaning toward having a single overload with default params so you can mix and match as necessary.  Seem good?